File: src/Java/BinarySearch/BinarySearchExample.java
package Java.BinarySearch;

// Simple BinarySearch Algorithm implementation.

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

import java.util.Arrays;

public class BinarySearchExample {
    public static int indexOf(int[] whitelist, int key) {
        // Return Index value of key value. (BinarySearch Algorithm)
        int lo = 0; // Starting Index (lo)
        int hi = whitelist.length - 1;  // Starting Index (hi)

        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (key < whitelist[mid]) {
                hi = mid - 1;   // Keep odd values of searching array.
            } else if (key > whitelist[mid]) {
                lo = mid + 1;   // Keep odd values of searching array.
            } else {
                return mid; // Return index of target value.
            }
        }
        return -1;
    }

    public static int indexOf(int[] whitelist, int lo, int hi, int key) {
        // Same BinarySearch Method, but using recursive.
        if (lo >= hi) {
            return  -1;
        } else {
            int mid = (lo + hi) / 2;
            if (key < whitelist[mid]) {
                hi = mid - 1;
                return indexOf(whitelist, lo, hi, key);
            } else if (key > whitelist[mid]) {
                lo = mid + 1;
                return indexOf(whitelist, lo, hi, key);
            } else {
                return mid; //Return index of target value.
            }
        }
    }

    // Development Client
    public static void main(String[] args) {
        StdOut.println("Start Development Client");
        In input = new In(args[0]);
        int[] whitelist = input.readAllInts();

        // Using BinarySearch Algorithm, array must be sorted in Ascending order.
        Arrays.sort(whitelist);

        while(!StdIn.isEmpty()) {
            int target = StdIn.readInt();
            if (indexOf(whitelist, target) == -1) {
                // If target is not in whitelist, print target.
                StdOut.println(target);
            }
        }
    }
}

File: src/Kotlin/BinarySearch/BinarySearchExample.kt
package Kotlin.BinarySearch

// Development Client
fun main() {
    var whitelist: List<Int> = listOf(84, 48, 68, 10, 18, 98, 12, 23, 54, 57, 33, 16, 77, 11, 29)
    var targetList: List<Int> = listOf(23, 50, 10, 99, 18, 23, 98, 84, 11, 10, 48, 77, 13, 54, 98, 77, 77, 68)

    var sortedWhiteList = whitelist.sorted()

    for (element in targetList) {
        if (indexOf(sortedWhiteList, element) == -1) {
            println(element)
        }
    }

    println("Binary Search Algorithm using recursive.")
    for (element in targetList) {
        if (indexOf(sortedWhiteList, 0, sortedWhiteList.size - 1, element) == -1) {
            println(element)
        }
    }

}

fun indexOf(array: List<Int>, key: Int) : Int {
    var lo = 0
    var hi = array.size - 1

    while (lo <= hi) {
        var mid = (lo + hi) / 2
        if (key < array[mid]) {
            hi = mid - 1;
        } else if (key > array[mid]) {
            lo = mid + 1;
        } else {
            return mid;
        }
    }
    return -1
}

fun indexOf(array: List<Int>, lo: Int, hi: Int, key: Int): Int {
    if (lo > hi) {
        return -1;
    } else {
        var mid = (lo + hi) / 2
        if (key < array[mid]) {
            // hi = mid - 1
            return indexOf(array, lo, mid - 1, key)
        } else if (key > array[mid]) {
            // lo = mid + 1
            return indexOf(array, mid + 1, hi, key)
        } else {
            return mid
        }
    }
}